sw/device/tests/sim_dv/alert_handler_escalation.c

```diff
@@ -0,0 +1,162 @@
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/base/math.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_alert_handler.h"
+#include "sw/device/lib/dif/dif_clkmgr.h"
+#include "sw/device/lib/dif/dif_keymgr.h"
+#include "sw/device/lib/dif/dif_rstmgr.h"
+#include "sw/device/lib/dif/dif_rv_core_ibex.h"
+#include "sw/device/lib/dif/dif_uart.h"
+#include "sw/device/lib/irq.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/alert_handler_testutils.h"
+#include "sw/device/lib/testing/keymgr_testutils.h"
+#include "sw/device/lib/testing/test_framework/FreeRTOSConfig.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+
+#include "alert_handler_regs.h"  // Generated.
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "rv_core_ibex_regs.h"  // Generated.
+
+/*
+  - Verify the first escalation results in NMI interrupt serviced by the CPU.
+  - Verify the second results in device being put in escalate state, via the LC
+  JTAG TAP.
+  - Verify the third results in chip reset.
+  - Ensure that all escalation handshakes complete without errors.
+
+  The first escalation is checked via the entry of the NMI handler and polling
+  by dv. The second escalation is directly checked by dv. The third escalation
+  is checked via reset reason.
+ */
+
+OTTF_DEFINE_TEST_CONFIG();
+
+static dif_clkmgr_t clkmgr;
+static dif_keymgr_t keymgr;
+static dif_rstmgr_t rstmgr;
+static dif_alert_handler_t alert_handler;
+static dif_rv_core_ibex_t rv_core_ibex;
+static dif_uart_t uart;
+
+typedef struct node {
+  const char *name;
+  dif_alert_handler_alert_t alert;
+  dif_alert_handler_class_t class;
+} node_t;
+
+static volatile uint32_t global_alert_called;
+static const dif_alert_handler_escalation_phase_t kEscProfiles[] = {
+    {.phase = kDifAlertHandlerClassStatePhase0,
+     .signal = 0,
+     .duration_cycles = 5000},
+    // TODO:
+    // this second duration must be long enough to
+    // accommodate a jtag transaction
+    // how can this be done in a non-hardcoded way?
+    {.phase = kDifAlertHandlerClassStatePhase1,
+     .signal = 1,
+     .duration_cycles = 10000},
+    {.phase = kDifAlertHandlerClassStatePhase2,
+     .signal = 3,
+     .duration_cycles = 3000}};
+
+static const dif_alert_handler_class_config_t kConfigProfiles[] = {{
+    .auto_lock_accumulation_counter = kDifToggleDisabled,
+    .accumulator_threshold = 0,
+    .irq_deadline_cycles = 0,
+    .escalation_phases = kEscProfiles,
+    .escalation_phases_len = 3,
+    .crashdump_escalation_phase = kDifAlertHandlerClassStatePhase1,
+}};
+
+/**
+ * External ISR.
+ *
+ * Handles all peripheral interrupts on Ibex. PLIC asserts an external interrupt
+ * line to the CPU, which results in a call to this OTTF ISR. This ISR
+ * overrides the default OTTF implementation.
+ */
+void ottf_external_nmi_handler(void) {
+  LOG_INFO("You are experiencing an NMI");
```

Comment:
Put a comment as in line 138 for safety.

Reply:
added